Atru/mcp server 1.0 ga (#2566)
hashicorp/web-unified-docs: content/terraform-mcp-server/v1.0.x/docs/mcp-server/deploy/remote.mdx

---
page_title: Deploy Terraform MCP server to a remote environment 

> description:

Reviewer: ```suggestion
description: Learn how to deploy Terraform MCP server to a remote environment and integrate with your agent.
```

---

# Deploy Terraform MCP server to a remote environment 

 
This topic describes how to deploy the Terraform Model Context Protocol (MCP) server into a remote environment so that you can centrally manage an instance of the server. This ensures that team members always have access to the same set of tools and that your organization can uniformly enforce security policies for accessing your registry and other platform resources.  
 

> For information about deploying the server locally, refer to [Deploy Terraform MCP server locallly](/terraform/mcp-server/deploy/local). 

Reviewer: ```suggestion
For information about deploying the server locally, refer to [Deploy Terraform MCP server locally](/terraform/mcp-server/deploy/local). 
```

## Overview 
 

> Choose the environment for your deployment. You can deploy the server into several types of environments, such as running the compiled binary in a cloud instance and running the containerized image in Docker, Fargate, or other container. You can also install the server from source if you intend to develop your own tools or customizations.  

Reviewer: ```suggestion
Choose the environment for your deployment. You can deploy the server into several types of environments, such as running the compiled binary in a cloud instance or running the containerized image in Docker, Fargate, or other container. You can also install the server from source if you intend to develop your own tools or customizations.  
```

I _think_ this sentence is trying to show lots of optionality, and the suggestion is trying to reinforce that.

 
If you intend to connect the server to your HCP Terraform or Terraform Enterprise account, create an API token for authentication.  
 
Configure the server. The only requirement for deploying the server to a remote environment is setting the transport mode to `streamable-http`, but you can configure any additional settings, such as enabling metrics, that match your requirements. To connect to HCP Terraform or Terraform Enterprise, you must also configure the server to present your API token.  
 
Configure your agent to connect to the remote server so that when a prompt triggers the Terraform MCP server, your local agent routes the request to the remote server. 

## Requirements 

The requirements depend on your installation environment. 
 
### Docker container 

> To use the Terraform HCP server Docker image, you must have Docker Engine v20.10.21+ or Docker Desktop v4.14.0+ installed on the local machine. Refer to the [Docker documentation](https://docs.docker.com/desktop) for installation instructions. 

Reviewer: ```suggestion
To use the Terraform MCP server Docker image, you must have Docker Engine v20.10.21+ or Docker Desktop v4.14.0+ installed on the target machine. Refer to the [Docker documentation](https://docs.docker.com/desktop) for installation instructions. 
```

### Pre-compiled binary 

You can run the compiled binary in the following operating systems and architectures: 
 
- MacOS: amd64, arm64 
- FreeBSD:  386, amd64 
- Linux: 386, amd64, arm, arm64 
- Solaris: amd64 
- Windows: 386, amd64  

 
### Source 

> You must install the [Go development environment](https://go.dev/).  

Reviewer: Wouldn't you install Go dev locally, build, and then deploy the binary elsewhere?

Author: Good question. @vravind1 , what is the expected workflow here? I was thinking (but it's not clear here) that you would install Go on the remote instance and build it there. Or are we suggesting that you build the binary locally, then deploy it to the instance, which seems like almost the same thing as just downloading a pre-compiled binary, then deploying it to the instance.

Reviewer 2: you dont need go dev here, since you are deploying the container in this section

 
Refer to the documentation for your specific deployment environment for additional information on how to pass environment variables. 

### Enable telemetry

It is optional, but you can configure Terraform MCP server to collect telemetry metrics. Set the `OTEL_METRICS_ENABLED` environment variable to `true` to enable metrics. The server implements the following lifecycle:

1. Creates an OTel meter provider 
1. Exports metrics over `OTLP/HTTP` to `OTEL_METRICS_ENDPOINT` 
1. Flushes the metrics according to the interval set by `OTEL_METRICS_EXPORT_INTERVAL` 
1. Adds the service identity is `OTEL_METRICS_SERVICE_NAME` and `OTEL_METRICS_SERVICE_VERSION`

Terraform MCP server generates the metrics and exports them to the configured OTLP endpoint. Then, the server can receive and forward the telemetry to Datadog through its own collector or forwarding setup. This repository produces OTel metrics, but the remote repository is responsible for routing them to Datadog.

Refer to [telemetry metrics reference](/terraform/mcp-server/reference#telemetry-metrics) for more information.

> ### Deploy the server

Reviewer 2: I think we’re missing a critical piece of functionality around per-request credential handling.
The server supports passing TFE_TOKEN via HTTP headers, but this isn’t currently documented. Without this, the feature appears to only support a single shared token, which hides an important capability: per-user RBAC via header override. It would be helpful to include a short example showing how to set these headers in an MCP client configuration.

Reviewer 2: Also, it would be helpful to clarify how credentials are resolved when both environment variables and request headers are present.
For example, noting that request headers take precedence over server environment variables would make the behavior clearer, especially for centralized or multi-user deployments.

Author: Sorry, I'm a little confused -- do you mean that when you deploy the server, you can include a token that has greater (or lesser) permissions to determine the scope of access? Or do you mean that _after_ deploying the server, you can present a different token than the one it's configured with and the server will present the token you provide when prompting your model? In either case, how does that work? Can you propose some content that we can add? If you mean the second use case, I think that should appear in the Prompt topic because this page is only about deploying the server.

Reviewer 2: I’m referring to the second case.
After the server is deployed (with or without a baseline token), a client can send a TFE_TOKEN via HTTP headers on each request. When present, the server uses that token instead of the environment variable, allowing different clients to access resources based on their own RBAC permissions.
This enables a centralized MCP server to serve multiple users while preserving per-user access control.
I think we could capture this with something like:

When using the HTTP transport, clients can provide a `TFE_TOKEN` via HTTP request headers:

```json
{
  "servers": {
    "terraform": {
      "type": "http",
      "url": "http://localhost:8080/mcp",
      "headers": {
        "TFE_TOKEN": "${input:tfe-token}"
      }
    }
  }
}
```

When a TFE_TOKEN is provided in the request headers, it takes precedence over any token configured on the server via environment variables. This allows a single MCP server deployment to support multiple users with different access permissions.

Reviewer 3: To expand a bit on this, this is a similar setup to the tutorial we have ready to review. We actually omit the server token altogether (to keep the tutorial as simple as possible), but in any case it seems the MCP server will opt to use a token included in the `TFE_TOKEN` header over any locally configured token.

For example, I might configure the server with a token that only has permission to read from the registry, but not create workspaces/run Terraform operations. Then, the user can provide their own token with more permissions to enable these extra abilities.

        logDriver = "awslogs"  
        options = {  
          "awslogs-group"         = aws_cloudwatch_log_group.mcp.name  
          "awslogs-region"        = var.region  
          "awslogs-stream-prefix" = "ecs"  
        }  
      }  
    }  
  ])  
}  
   
resource "aws_ecs_service" "mcp" {  
  name            = "terraform-mcp-${random_string.suffix.id}"  
  cluster         = aws_ecs_cluster.compute.id  
  task_definition = aws_ecs_task_definition.mcp.arn  
  desired_count   = 1  
  launch_type     = "FARGATE"  
   
  network_configuration {  
    subnets          = module.vpc.private_subnets  
    security_groups  = [aws_security_group.fargate_traffic.id]  
    assign_public_ip = false  
  }  
   
  load_balancer {  
    target_group_arn = aws_lb_target_group.mcp.arn  
    container_name   = "terraform-mcp-${random_string.suffix.id}"  
    container_port   = var.port  
  }  
   
  depends_on = [aws_lb.mcp]  
}  
   
resource "aws_cloudwatch_log_group" "mcp" {  
  name              = "/ecs/terraform-mcp"  
  retention_in_days = 1  
}  
```
  
</CodeBlockConfig>

## Configure local agents 

Add the URL of the remote server to your agent’s MCP configuration file. 

> The following example shows an entry in the `.vscode/mcp.json` file for Visual Studio Code. Refer to your agent documentation for instructions on how to specify a remote MCP server:   

Reviewer: I think this is a very reasonable approach.

---
page_title: Deploy Terraform MCP server to a remote environment 
description: Learn how to deploy Terraform MCP server to a remote environment and integrate with your agent.

---

# Deploy Terraform MCP server to a remote environment 

 
This topic describes how to deploy the Terraform Model Context Protocol (MCP) server into a remote environment so that you can centrally manage an instance of the server. This ensures that team members always have access to the same set of tools and that your organization can uniformly enforce security policies for accessing your registry and other platform resources.  
 
For information about deploying the server locally, refer to [Deploy Terraform MCP server locally](/terraform/mcp-server/deploy/local). 

## Overview 
 

> Choose the environment for your deployment. You can deploy the server into several types of environments, such as running the compiled binary in a cloud instance or running the containerized image in Docker, Fargate, or other container. 

Reviewer 3: ```suggestion
Choose the environment for your deployment. You can deploy the server into several types of environments, such as running the compiled binary in a cloud instance or running the containerized image in Docker, Fargate, or other container engine. 
```

> If you intend to connect the server to your HCP Terraform or Terraform Enterprise account, create an API token for authentication.  

Reviewer 3: ```suggestion
If you intend to connect the server to your HCP Terraform or Terraform Enterprise organization, create an API token for authentication.  
```

Similar reasoning as the local deployment docs

 
Configure the server. The only requirement for deploying the server to a remote environment is setting the transport mode to `streamable-http`, but you can configure any additional settings, such as enabling metrics, that match your requirements. To connect to HCP Terraform or Terraform Enterprise, you must also configure the server to present your API token.  
 
Configure your agent to connect to the remote server so that when a prompt triggers the Terraform MCP server, your local agent routes the request to the remote server. 

## Requirements 

The requirements depend on your installation environment. 
 
### Docker container 

To use the Terraform MCP server Docker image, you must have Docker Engine v20.10.21+ or Docker Desktop v4.14.0+ installed on the target machine. Refer to the [Docker documentation](https://docs.docker.com/desktop) for installation instructions. 

> 

Reviewer 3: Is this still relevant for the remote docs? I suppose it's applicable if you are running Docker engine on a server, however we also mention running the container on Fargate/ECS for example.

 
### HCP Terraform and Terraform Enterprise 

If connecting to HCP Terraform or Terraform Enterprise, you must have network access to `https://app.terraform.io` or your Terraform Enterprise endpoint. 
 
You must also provide the server with an API token for accessing HCP Terraform or Terraform Enterprise. Refer to [API Tokens](/terraform/cloud-docs/users-teams-organizations/api-tokens) in the HCP Terraform documentation or [API Tokens](/terraform/enterprise/users-teams-organizations/api-tokens) in the Terraform Enterprise documentation for instructions on creating API tokens. 

## Configure the server 

Set environment variables to define server behavior and pass them into your deployment environment. You must set the `TRANSPORT_MODE` variable to `streamable-http` so that the server receives and sends data using a single endpoint. Refer to the [environment variables reference documentation](/terraform/mcp-server/reference#environment-variables) for information about all available settings.  
 
Refer to the documentation for your specific deployment environment for additional information on how to pass environment variables. 

### Connect to HCP Terraform or Terraform Enterprise 

> You must configure the server with a baseline API token for authentication by setting the `TFE_TOKEN` variable and the HCP Terraform or Terraform Enteprise instance by setting the `TFE_ADDRESS` variable. Note that the server supports presenting multiple tokens using HTTP headers. Refer to [Configure local agents](#configure-local-agents) for more information.

Reviewer 3: ```suggestion
You must configure the server with a baseline API token for authentication by setting the `TFE_TOKEN` variable to your API token. Note that the server supports presenting multiple tokens using HTTP headers. Refer to [Configure local agents](#configure-local-agents) for more information.

To connect the server to Terraform Enterprise, you must also set the `TFE_HOSTNAME` environment variable to the URL of your Terraform Enterprise deployment.
```

> You must configure the server with a baseline API token for authentication by setting the `TFE_TOKEN` variable and the HCP Terraform or Terraform Enteprise instance by setting the `TFE_ADDRESS` variable. Note that the server supports presenting multiple tokens using HTTP headers. Refer to [Configure local agents](#configure-local-agents) for more information.

Reviewer 4: It's optional to configure the `TFE_TOKEN` when you start the server. You can start the server without it and still supply the token via HTTP headers.

### Enable telemetry

> It is optional, but you can configure Terraform MCP server to collect telemetry metrics. Set the `OTEL_METRICS_ENABLED` environment variable to `true` to enable metrics. The server implements the following lifecycle:

Reviewer 3: ```suggestion
It is optional, but you can configure the Terraform MCP server to collect telemetry metrics. Set the `OTEL_METRICS_ENABLED` environment variable to `true` to enable metrics. The server implements the following lifecycle:
```

1. Creates an OTel meter provider 
1. Exports metrics over `OTLP/HTTP` to `OTEL_METRICS_ENDPOINT` 
1. Flushes the metrics according to the interval set by `OTEL_METRICS_EXPORT_INTERVAL` 
1. Adds the service identity is `OTEL_METRICS_SERVICE_NAME` and `OTEL_METRICS_SERVICE_VERSION`

> Terraform MCP server generates the metrics and exports them to the configured OTLP endpoint. Then, the server can receive and forward the telemetry to Datadog through its own collector or forwarding setup. This repository produces OTel metrics, but the remote repository is responsible for routing them to Datadog.

Reviewer 3: ```suggestion
The Terraform MCP server generates the metrics and exports them to the configured OTLP endpoint. Then, the server can receive and forward the telemetry to your monitoring software through its own collector or forwarding setup. This repository produces OTel metrics, but the remote repository is responsible for routing them to the monitoring service.
```

Is Datadog the only observability service supported? I'd expect since the MCP server uses OpenTelemetry, I could forward these metrics to any service that supports that format. If it's not the only supported service, should we either explicitly say this is an example implementation, or at least make this paragraph more generic?

> Terraform MCP server generates the metrics and exports them to the configured OTLP endpoint. Then, the server can receive and forward the telemetry to Datadog through its own collector or forwarding setup. This repository produces OTel metrics, but the remote repository is responsible for routing them to Datadog.

Reviewer 4: This mentions Datadog explicitly – OpenTelemetry is vendor agnostic so we should remove this, customers can export to whatever monitoring stack supports OTLP. Datadog is what we were using internally so prob made it into a README somewhere.

        {  
          name  = "TRANSPORT_MODE"  
          value = "streamable-http"  
        },  
        {  
          name  = "TRANSPORT_HOST"  
          value = "0.0.0.0"  
        },  
        {  
          name  = "TRANSPORT_PORT"  
          value = tostring(var.port)  
        },  
        {  
          name  = "LOG_LEVEL"  
          value = var.log_level  
        },  
        {  
          name  = "TFE_ADDRESS",  
          value = var.tfe_address  
        },  
        {  
          name  = "TFE_TOKEN",  
          value = var.tfe_token  
        }  
      ]  
   
      portMappings = [{  
        containerPort = var.port  
        protocol      = "tcp"  
      }]  
   
      logConfiguration = {  
        logDriver = "awslogs"  
        options = {  
          "awslogs-group"         = aws_cloudwatch_log_group.mcp.name  
          "awslogs-region"        = var.region  
          "awslogs-stream-prefix" = "ecs"  
        }  
      }  
    }  
  ])  
}  
   
resource "aws_ecs_service" "mcp" {  
  name            = "terraform-mcp-${random_string.suffix.id}"  
  cluster         = aws_ecs_cluster.compute.id  
  task_definition = aws_ecs_task_definition.mcp.arn  

>   desired_count   = 1  

Reviewer 5: Worth a note here for anyone scaling past one replica. The default stateful mode keeps session state in each instance's memory, so with multiple replicas behind a load balancer, a client's initialize can hit one replica and its next tools/call another that doesn't know the session. The README already points to stateless for load-balanced/HA — I think a quick line telling multi-replica deployments to set MCP_SESSION_MODE=stateless would save someone any potential issues

  launch_type     = "FARGATE"  
   
  network_configuration {  
    subnets          = module.vpc.private_subnets  
    security_groups  = [aws_security_group.fargate_traffic.id]  
    assign_public_ip = false  
  }  
   
  load_balancer {  
    target_group_arn = aws_lb_target_group.mcp.arn  
    container_name   = "terraform-mcp-${random_string.suffix.id}"  
    container_port   = var.port  
  }  
   
  depends_on = [aws_lb.mcp]  
}  
   
resource "aws_cloudwatch_log_group" "mcp" {  
  name              = "/ecs/terraform-mcp"  
  retention_in_days = 1  
}  
```
  
</CodeBlockConfig>

## Configure local agents 

Add the URL of the remote server to your agent’s MCP configuration file. 

The following example shows an entry in the `.vscode/mcp.json` file for Visual Studio Code. Refer to your agent documentation for instructions on how to specify a remote MCP server:   
 
```json 
“mcpServers”: { 
  "remote-terraform-mcp-server": { 
     "url": “http://remote-server-path:port/endpoint”  
     "type": "http" 
  } 

> } 

Reviewer 3: ```suggestion
{
  "servers": {
    "remote-terraform-mcp-server": {
      "url": "http://remote-server-path:port/endpoint",
      "type": "http"
    }
  }
}
```

Updating to match a couple things from the config I've been using:

1. Fixed unicode quotation marks to ASCII
2. VS Code expects `servers` as the top-level name for the list of MCP servers ([ref](https://code.visualstudio.com/docs/agents/reference/mcp-configuration#_configuration-structure))
3. VS Code expects the entire config to be one single object

```

### Configure a user token

> Instead of using the baseline token the server is configured to present, you can also configure your client to pass your user API token to the remote server for authentication with HCP Terraform or Terraform Enterprise. This lets different clients access resources based on their RBAC permissions so that mutliple users with different access levels can access data they need using a centrally managed MCP server.  

Reviewer 3: ```suggestion
Instead of using the baseline token the server is configured to present, you can also configure your client to pass your API token to the remote server for authentication with HCP Terraform or Terraform Enterprise. This lets different clients access resources based on their RBAC permissions so that multiple users with different access levels can access data they need using a centrally managed MCP server.  
```

Removing the specific callout to a user token as they can also use team tokens (which would actually be recommended IMO), org tokens, etc

Configure your client to pass the user API token in the HTTP header, for example:

```json

> {
>   "servers": {
>     "terraform": {
>       "type": "http",
>       "url": "http://localhost:8080/mcp",
>       "headers": {
>         "TFE_TOKEN": "${input:tfe-token}"
>       }
>     }
>   }
> }

Reviewer 3: ```suggestion
{
  "servers": {
    "remote-terraform-mcp-server": {
      "url": "http://remote-server-path:port/endpoint",
      "type": "http",
      "headers": {
        "TFE_TOKEN": "${input:tfe-token}"
      }
    }
  }
}
```

Small nit but brought more in line to mirror the above example
